src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -861,6 +861,84 @@ void StorageReplicatedMergeTree::createReplica(const StorageMetadataPtr & metada
+    const String local_columns = metadata_snapshot->getColumns().toString();
+    const String local_metadata_version = toString(metadata_snapshot->getMetadataVersion());
+
+    /// It is possible for the replica to fail after creating ZK nodes without saving local metadata.
+    /// Because of that we need to check whether the replica exists and is newly created.
+    /// For this we check that all nodes exist, the metadata of the table is the same, and other nodes are not modified.
+
+    std::vector<String> paths_exists = {
+        replica_path,
+        replica_path + "/host",
+        replica_path + "/log_pointer",
+        replica_path + "/queue",
+        replica_path + "/parts",
+        replica_path + "/flags",
+        replica_path + "/is_lost",
+        replica_path + "/metadata",
+        replica_path + "/columns",
+        replica_path + "/metadata_version",
+        replica_path + "/mutation_pointer",
+        replica_path + "/min_unprocessed_insert_time",
+        replica_path + "/max_processed_insert_time",
+        replica_path + "/mutation_pointer"
+    };
+
+    auto response_exists = zookeeper->tryGet(paths_exists);
+    size_t response_num = 0;
+
+    if (response_exists[response_num++].error == Coordination::Error::ZOK)
+    {
+        bool all_nodes_exist = true;
+
+        for (size_t i = 0; i < response_exists.size(); ++i)
+        {
+            if (response_exists[i].error != Coordination::Error::ZOK)
+            {
+                all_nodes_exist = false;
+                break;
+            }
+        }
+
+        if (all_nodes_exist)
+        {
+            const auto & zk_host                        = response_exists[response_num++].data;
+            const auto & zk_log_pointer                 = response_exists[response_num++].data;
+            const auto & zk_queue                       = response_exists[response_num++].data;
+            const auto & zk_parts                       = response_exists[response_num++].data;
+            const auto & zk_flags                       = response_exists[response_num++].data;
+            const auto & zk_is_lost                     = response_exists[response_num++].data;
+            const auto & zk_metadata                    = response_exists[response_num++].data;
+            const auto & zk_columns                     = response_exists[response_num++].data;
+            const auto & zk_metadata_version            = response_exists[response_num++].data;
+            const auto & zk_min_unprocessed_insert_time = response_exists[response_num++].data;
+            const auto & zk_max_processed_insert_time   = response_exists[response_num++].data;
+            const auto & zk_mutation_pointer            = response_exists[response_num++].data;
+
+            if (zk_host.empty() &&
+                zk_log_pointer.empty() &&
+                zk_queue.empty() &&
+                zk_parts.empty() &&
+                zk_flags.empty() &&
```

Comment:
`is_lost_value` is always empty.
`zk_queue` is always empty too, but it may have children

Reply:
The main point for these two and the `is_lost_value` checks is to make sure that the created nodes in the ZooKeeper are the same as the ones we would create otherwise. Also I don't think that it makes sense to check the children of the `queue`, since at that point we would surely change something else.

Reply:
> since at that point we would surely change something else.

I would not be that sure, but yes, we update some of these nodes in early stages of initialization, for example https://github.com/ClickHouse/ClickHouse/blob/0502a4c73e11f4377ead87f00a60d6806e15d489/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L286-L299